### PR TITLE
use subcommands in plugin updaters

### PR DESCRIPTION
**Pull request by teto**
- update.py: introduce subparsers for plugin updaters
- doc/updaters: update command to update editor plugins

###### Description of changes

This is preliminary work to help create more powerful plugin updaters.
Namely I would like to be able to "just add" plugins without refreshing
the older ones (helpful when github temporarily removes a user from
github due to automated bot detection).

Also concerning the lua updater, we pin some of the dependencies, and I
would like to be able to unpin the package without editing the csv
(coming in later PRs).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by mrcjkb — APPROVED** [review]

File: doc/languages-frameworks/vim.section.md [hunk 1]
```diff
@@ -236,12 +236,13 @@ Run the update script with a GitHub API token that has at least `public_repo` ac
 
 ```sh
 GITHUB_API_TOKEN=my_token ./pkgs/applications/editors/vim/plugins/update.py
+update
```
Comment (figsoda) [inline doc/languages-frameworks/vim.section.md:239]: can we make `update` the default subcommand?

File: doc/languages-frameworks/vim.section.md [hunk 2]
```diff
@@ -212,7 +212,7 @@ Note: this is not possible anymore for Neovim.
 
 ## Adding new plugins to nixpkgs {#adding-new-plugins-to-nixpkgs}
 
-Nix expressions for Vim plugins are stored in [pkgs/applications/editors/vim/plugins](https://github.com/NixOS/nixpkgs/tree/master/pkgs/applications/editors/vim/plugins). For the vast majority of plugins, Nix expressions are automatically generated by running [`./update.py`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vim/plugins/update.py). This creates a [generated.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vim/plugins/generated.nix) file based on the plugins listed in [vim-plugin-names](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vim/plugins/vim-plugin-names). Plugins are listed in alphabetical order in `vim-plugin-names` using the format `[github username]/[repository]@[gitref]`. For example https://github.com/scrooloose/nerdtree becomes `scrooloose/nerdtree`.
```
Comment (figsoda) [inline doc/languages-frameworks/vim.section.md:215]: why is this removed?

File: maintainers/scripts/pluginupdate.py [hunk 3]
```diff
@@ -395,59 +426,97 @@ def rewrite_input(self, *args, **kwargs):
...
+
+        # update only a subset ? --all ?
+        pupdate = subparsers.add_parser(
+            "update", parents=[common],
+            # parents= just copies arguments, not the actual help !
+            description="Update all or a subset of existing plugins",
+            add_help=False,
+        )
+        pupdate.set_defaults(func=self.update)
+        return main
```
[the remark this answers is not in the document]
Comment (teto) [reply maintainers/scripts/pluginupdate.py:505]: the script is run whenever someone wants to add basic (neo)vim plugin or just want to update the plugins. If treesitter is updated, it updates the grammar to match the lockfile

[the remark this answers is not in the document]
Comment (teto) [reply maintainers/scripts/pluginupdate.py:505]: the goal is to make the script more robust, e.g. right now it always need to update the plugins, which sometimes fail. I wish we could just keep the file as is and just add a plugin, hence the PR.

File: doc/languages-frameworks/vim.section.md (hunk 1 above)
Comment (teto) [reply doc/languages-frameworks/vim.section.md:239]: I guess we could (not requiring a subparser) but I am not keen of that, I feel like it would defeat the purpose of the PR to make a cleaner distinction. 
In case someone doesn't mention, the error looks descriptive enough
```
$ pkgs/applications/editors/vim/plugins/update.py
usage: update.py [-h] {add,update} ...
update.py: error: the following arguments are required: command
```

File: doc/languages-frameworks/vim.section.md (hunk 2 above)
Comment (teto) [reply doc/languages-frameworks/vim.section.md:215]: because it was outdated, the format is now csv with full uris. I dont think it's worth mentioning in the doc anyway, it is obvious once you want to add a plugin.

File: doc/languages-frameworks/vim.section.md (hunk 1 above)
Comment (figsoda) [reply doc/languages-frameworks/vim.section.md:239]: I would prefer it to be the default since the file name is already `update.py`

Comment (mrcjkb) [reply doc/languages-frameworks/vim.section.md:239]: Would it make sense to split these into separate `update.py` and `add.py` scripts?

File: maintainers/scripts/pluginupdate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrcjkb) [reply maintainers/scripts/pluginupdate.py:505]: I understood that. I was just wondering how frequently someone runs `update` without adding new plugins, because I couldn't find a scheduled workflow for doing so.

File: doc/languages-frameworks/vim.section.md (hunk 1 above)
Comment (teto) [reply doc/languages-frameworks/vim.section.md:239]: > Would it make sense to split these into separate update.py and add.py scripts

I dont think so, it would hurt discoverability IMO. 

> I would prefer it to be the default since the file name is already update.py

if you add a plugin, you update nixpkgs too. I would prefer `updater` as a name, e.g. `pkgs/applications/editors/vim/plugins/updater` but I did not want to make this more confusing. 
At first I thought I would add subcommand-specific flags like `--unpin <PKG>` but if we update by default it might be better to have `unpin` as its own command (alternatively update --force <PKG> ? well that can be discussed in a later PR).

File: maintainers/scripts/pluginupdate.py [hunk 4]
```diff
@@ -344,12 +344,43 @@ def __init__(
...
+        fetch_config = FetchConfig(args.proc, args.github_token)
+        editor = self
+        # TODO moved to add
+        for plugin_line in args.add_plugins:
+            print("using plugin_line", plugin_line)
+            pdesc = PluginDesc.load_from_string(fetch_config, plugin_line)
+            print(pdesc)
+            append = [ pdesc ]
+            editor.rewrite_input(fetch_config, args.input_file, editor.deprecated, append=append)
+            # update()
```
[the remark this answers is not in the document]
Comment (teto) [reply maintainers/scripts/pluginupdate.py]: done

File: doc/languages-frameworks/vim.section.md (hunk 1 above)
Comment (teto) [reply doc/languages-frameworks/vim.section.md]: I've made `update` the default and reverted the related doc changes accordingly

**Review by figsoda — COMMENTED** [review]
`--github-token` doesn't seem to work anymore, at least for the vim update script

**Comment by teto** [comment]
@figsoda this should be fixed. To sum up, the only effective change right now is instead do doing `update.py --add XXX` is that you should now run `update.py add`. The endgame is more ambitious but let's break stuff after the nixos release and not right before. Also I can't help but feel like we are recreating some abstraction over lockfiles. I feel like we should do it differently (more integration with nix-update or flakes or I dunno) but this should be adressed later.

**Comment by figsoda** [comment]
thanks, I will test it out in a bit

File: pkgs/applications/editors/vim/plugins/update.py [hunk 5]
```diff
@@ -42,6 +42,9 @@
 import pluginupdate
 from pluginupdate import run_nix_expr, PluginDesc
 
+
+
+# pkgs/applications/editors/vim/plugins/get-plugins.nix
 GET_PLUGINS = f"""(with import <localpkgs> {{}};
```
Comment (figsoda) [inline pkgs/applications/editors/vim/plugins/update.py:48]: this seems to be unused now?

File: pkgs/applications/editors/vim/plugins/get-plugins.nix [hunk 6]
```diff
@@ -0,0 +1,20 @@
+with import <localpkgs> {};
+let
+  inherit (vimUtils.override {inherit vim;}) buildVimPluginFrom2Nix;
+  inherit (neovimUtils) buildNeovimPluginFrom2Nix;
+
+  generated = callPackage ./pkgs/applications/editors/vim/plugins/generated.nix {
```
Comment (figsoda) [inline pkgs/applications/editors/vim/plugins/get-plugins.nix:6]: would this work when you are `cd`ed into the dirctory?

File: pkgs/applications/editors/vim/plugins/update.py [hunk 7]
```diff
@@ -118,34 +121,39 @@ def plugin2nix(self, pdesc: PluginDesc, plugin: pluginupdate.Plugin) -> str:
...
+                print("no updates to nvim-treesitter grammars")
+
+
 def main():
 
     global luaPlugins
     luaPlugins = run_nix_expr(GET_PLUGINS_LUA)
 
+    with open("pkgs/applications/editors/vim/plugins/get-plugins.nix") as f:
+        GET_PLUGINS = f.read()
```
Comment (figsoda) [inline pkgs/applications/editors/vim/plugins/update.py:154]: is this required for the change? this looks like an unrelated refactor to me

File: pkgs/applications/editors/vim/plugins/update.py (hunk 5 above)
Comment (teto) [reply pkgs/applications/editors/vim/plugins/update.py:48]: indeed I removed it

File: pkgs/applications/editors/vim/plugins/get-plugins.nix (hunk 6 above)
Comment (teto) [reply pkgs/applications/editors/vim/plugins/get-plugins.nix:6]: it didn't I just fixed it.

File: pkgs/applications/editors/vim/plugins/update.py (hunk 7 above)
Comment (teto) [reply pkgs/applications/editors/vim/plugins/update.py:154]: It's kinda unrelated but this is how I realized the caching mechanism wasn't working anymore: having the GET_PLUGINS in its own file, I was able to load it in a REPL and iterate over it quickly.

File: pkgs/applications/editors/vim/plugins/get-plugins.nix [hunk 8]
```diff
@@ -0,0 +1,20 @@
+with import <localpkgs> {};
+let
+  inherit (vimUtils.override {inherit vim;}) buildVimPluginFrom2Nix;
+  inherit (neovimUtils) buildNeovimPluginFrom2Nix;
+
+  generated = callPackage <localpkgs/pkgs/applications/editors/vim/plugins/generated.nix> {
+    inherit buildNeovimPluginFrom2Nix buildVimPluginFrom2Nix;
+  } {} {};
+  hasChecksum = value:
+    # true;
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vim/plugins/get-plugins.nix:10]:
```suggestion
```

**Review by figsoda — APPROVED** [review]
